src/waiter.c

```diff
@@ -0,0 +1,311 @@
+                        waiter->notification_callback = notification_callback;
+                        waiter->notification_callback_context = context;
+                        waiter->registered = true;
+                        try_complete(waiter);
+                        THANDLE_INITIALIZE_MOVE(ASYNC_OP)(&result, &async_op);
+                        success = true;
+                    }
+                }
+            }
+            srw_lock_release_exclusive(waiter->lock);
+
+            if (!success)
+            {
+                LogError("Failure in waiter_register_notification(waiter=%p, notification_callback=%p, context=%p)", waiter, notification_callback, context);
+            }
+            else
+            {
+                goto all_ok;
+            }
+            sm_exec_end(waiter->sm);
+        }
+    }
+all_ok:
+    return result;
+}
+
+static void cancel_notify(void* context)
+{
+    WAITER_HANDLE waiter = context;
+    srw_lock_acquire_exclusive(waiter->lock);
+    {
+        waiter->notify_complete_callback(waiter->notify_complete_callback_context, WAITER_RESULT_CANCELLED);
+        waiter->callbacks_called = true;
+        THANDLE_ASSIGN(RC_PTR)(&waiter->data, NULL);
+        sm_fault(waiter->sm);
+        sm_exec_end(waiter->sm);
+    }
+    srw_lock_release_exclusive(waiter->lock);
+}
+
+THANDLE(ASYNC_OP) waiter_notify(WAITER_HANDLE waiter, THANDLE(RC_PTR) data, NOTIFY_COMPLETE_CALLBACK notify_complete_callback, void* context)
+{
+    THANDLE(ASYNC_OP) result = NULL;
+    if (waiter == NULL || data == NULL || notify_complete_callback == NULL || context == NULL)
+    {
+        LogError("Invalid arguments: WAITER_HANDLE waiter = %p, THANDLE(RC_PTR) data = %p, NOTIFY_COMPLETE_CALLBACK notify_complete_callback = %p, void* context = %p",
+                   waiter, data, notify_complete_callback, context);
+    }
+    else
+    {
+        SM_RESULT sm_exec_begin_result = sm_exec_begin(waiter->sm);
+        if (sm_exec_begin_result != SM_EXEC_GRANTED)
+        {
+            LogError("Failure in sm_exec_begin(waiter->sm=%p)", waiter->sm);
+        }
+        else
+        {
+            bool success;
+            srw_lock_acquire_exclusive(waiter->lock);
+            {
+                if (waiter->notified)
+                {
+                    LogError("waiter_notify called twice on WAITER waiter=%p."
+                        "Previous arguments: THANDLE(RC_PTR) data=%" PRI_RC_PTR ", NOTIFY_COMPLETE_CALLBACK notify_complete_callback=%p, void* context=%p."
+                        "Current arguments: THANDLE(RC_PTR) data=%" PRI_RC_PTR ", NOTIFY_COMPLETE_CALLBACK notify_complete_callback=%p, void* context=%p.",
+                        waiter,
+                        waiter->data, waiter->notify_complete_callback, waiter->notify_complete_callback_context,
+                        data, notify_complete_callback, context
+                    );
+                    success = false;
```

Comment:
>false [](http://example.com/codeflow?start=30&length=5)

can this code be written without "success"?

Reply:
Yes, done in next iteration.